tinygrad/function.py

```diff
@@ -36,13 +36,371 @@ def forward(self, x:LazyBuffer) -> LazyBuffer:
+  def __lv1(x: LazyBuffer) -> LazyBuffer:
+    if fp32_p:
+      d = _mla(q, x.const(-3.1414794921875), x)
+      d = _mla(q, x.const(-0.00011315941810607910156), d)
+      d = _mla(q, x.const(-1.9841872589410058936e-09), d)
+    else:
+      d = _mla(q, x.const(-3.141592653589793116), x)
+      d = _mla(q, x.const(1.2246467991473532072e-16), d)
+    return d
+
+  def __lv2(x: LazyBuffer) -> LazyBuffer:
+    if fp32_p:
+      d = _mla(q, x.const(-3.1414794921875), x)
+      d = _mla(q, x.const(-0.00011315941810607910156), d)
+      d = _mla(q, x.const(-1.9841872589410058936e-09), d)
+      d = _mla(q, x.const(-1.2154201256553420762e-10), d)
+    else:
+      assert qdh is not None
+      d = _mla(qdh, x.const(-3.1415926218032836914), x)
+      d = _mla(q, x.const(-3.1415926218032836914), d)
+      d = _mla(qdh, x.const(-3.1786509424591713469e-08), d)
+      d = _mla(q, x.const(-3.1786509424591713469e-08), d)
+      d = _mla(qdh, x.const(-1.2246467864107188502e-16), d)
+      d = _mla(q, x.const(-1.2246467864107188502e-16), d)
+      d = _mla(qdh.e(BinaryOps.ADD, q), x.const(-1.2736634327021899816e-24), d)
+    return d
+
+  lv3_d = __lv2(d) if fast else __lv2(lv3_reduced_d)
+  d = di.e(BinaryOps.CMPLT, trig_range_lv1).e(TernaryOps.WHERE, __lv1(d), di.e(BinaryOps.CMPLT, trig_range_lv2).e(TernaryOps.WHERE, __lv2(d), lv3_d))
+  s = d.e(BinaryOps.MUL, d)
+  a = q.cast(dtypes.int64).e(BinaryOps.MOD, d.const(2).cast(dtypes.int64)).cast(d.dtype)
+  d = d.e(BinaryOps.MUL, a.e(BinaryOps.CMPNE, d.const(0)).e(TernaryOps.WHERE, d.const(-1), d.const(1)))
+
+  u = None
+  if fp32_p:
+    u = d.const(2.6083159809786593541503e-06)
+    u = _mla(u, s, u.const(-0.0001981069071916863322258))
+    u = _mla(u, s, u.const(0.00833307858556509017944336))
+    u = _mla(u, s, u.const(-0.166666597127914428710938))
+    u = _mla(s, u.e(BinaryOps.MUL, d), d)
+  else:
+    s2 = s.e(BinaryOps.MUL, s)
+    s4 = s2.e(BinaryOps.MUL, s2)
+
+    def __poly4(x: LazyBuffer, x2: LazyBuffer, c3, c2, c1, c0) -> LazyBuffer: return _mla(x2, _mla(x, d.const(c3), d.const(c2)), _mla(x, d.const(c1), d.const(c0))) # noqa: E501
+    def __poly8(x, x2, x4, c7, c6, c5, c4, c3, c2, c1, c0) -> LazyBuffer: return _mla(x4, __poly4(x, x2, c7, c6, c5, c4), __poly4(x, x2, c3, c2, c1, c0)) # noqa: E501
+    u = __poly8(
+      s,
+      s2,
+      s4,
+      -7.97255955009037868891952e-18,
+      2.81009972710863200091251e-15,
+      -7.64712219118158833288484e-13,
+      1.60590430605664501629054e-10,
+      -2.50521083763502045810755e-08,
+      2.75573192239198747630416e-06,
+      -0.000198412698412696162806809,
+      0.00833333333333332974823815,
+    )
+    u = _mla(u, s, d.const(-0.166666666666666657414808))
+    u = _mla(s, u.e(BinaryOps.MUL, d), d)
+  return u
+
+def _xsin(x: LazyBuffer, is_metal: bool=False, fast: bool=False) -> LazyBuffer:
```

Comment:
Hmm, why do you need an is_metal here? If metal is behaving differently we should probably fix it in the METAL backend if possible.

Reply:
When running in FP32, it was necessary to calculate the value of 2/pi in FP64 for the range where `x >= 39000.0`. However, since FP64 is not supported on METAL, I set is_metal=True to ignore fp64 computation.

Upon further investigation, I found that FP32 was sufficient, so I will remove `is_metal` later.

Reply:
I have deleted `is_metal` option.